cpmpy/transformations/linearize.py

```diff
@@ -5,190 +5,183 @@
+                return linearize_constraint(constraints, supported=supported)
+
+            # other global constraints
+            elif isinstance(lhs, GlobalConstraint) and hasattr(lhs, "decompose_comparison"):
+                decomp = lhs.decompose_comparison(cpm_expr.name, rhs)
+                return linearize_constraint(flatten_constraint(decomp),supported=supported)
 
-    gen_constr = ["min","max","abs","pow"] #General constraints supported by gurobi, TODO: make composable for other solvers
-    if cpm_expr.name == '==':
-        lhs, rhs = cpm_expr.args
-        if lhs.is_bool() and not isinstance(lhs, _BoolVarImpl) and isinstance(rhs, _BoolVarImpl):
-            # BE == BV :: ~BV -> ~BE, BV -> BE
-            if lhs.name in gen_constr:
-                return [cpm_expr]
             else:
-                c1 = (~rhs).implies(negated_normal(lhs))
-                c2 = rhs.implies(lhs)
-                return linearize_constraint(flatten_constraint([c1,c2]))
-        elif lhs.is_bool() and not isinstance(lhs, _BoolVarImpl) and isinstance(rhs,_NumVarImpl):
-            #boolexpr == intvar, so introduce an intermediate boolvar bv, boolexpr == bv, bv == intvar
-            bv = boolvar()
-            c1 = Comparison('==', lhs, bv)
-            c2 = Comparison('==', bv, rhs)
-            return linearize_constraint(flatten_constraint([c1, c2]))
-
-    if cpm_expr.name == "xor" and len(cpm_expr.args) == 2:
-        return [sum(cpm_expr.args) == 1]
+                raise NotImplementedError(f"lhs of constraint {cpm_expr} cannot be linearized, should be any of {supported} or 'sub', 'mul','element' but is {lhs}. Please report on github")
+
+
+
+
+        if isinstance(lhs, Operator) and lhs.name in {"sum","wsum"}:
+            # bring all vars to lhs
+            if isinstance(rhs, _NumVarImpl):
+                if lhs.name == "sum":
+                    lhs, rhs = sum([-1 * rhs]+[1 * a for a in lhs.args]), 0
+                else:
+                    lhs, rhs = lhs + -1*rhs, 0
+            # bring all const to rhs
+            if lhs.name == "sum":
+                rhs += -sum(arg for arg in lhs.args if is_num(arg))
+                lhs = sum(arg for arg in lhs.args if not is_num(arg)) # TODO: avoid looping again and keep idxes?
+            elif lhs.name == "wsum":
+                rhs += -sum(w * arg for w,arg in zip(*lhs.args) if is_num(arg))
+                lhs = sum(w * arg for w, arg in zip(*lhs.args) if not is_num(arg))
+
+        # now fix the comparisons themselves
+        if cpm_expr.name == "<":
+            new_rhs, cons = get_or_make_var(rhs - 1)
+            return [lhs <= new_rhs] + cons
+        if cpm_expr.name == ">":
+            new_rhs, cons = get_or_make_var(rhs + 1)
+            return [lhs >= new_rhs] + cons
+        if cpm_expr.name == "!=":
+            # Big-M implementation, ensure no new reifications are added
```

Comment:
- a new (smart) reification *is* added, see `z=boolvar()`.
- we count on solvers to do their own Big-M by allowing the `BoolVar -> LinExpr` output. We do this because we hope they can do it more efficiently. The same assumption should hold here, so we should just translate to that type of constraint.
- so just add `z -> linearize(lhs > rhs)` and `~z -> linearize(lhs < rhs)`?

Reply:
Also, this linearization will not work when the `LinExpr` is nested, e.g., `p -> x!=y`. Here we do need both reifications.

Reply:
Conclusion after discussion: current approach seems fine, except that we want to keep the newly introduced reification when not nested (as there is no risk of infinite recursive reifications).